core/shared/src/test/scala/fs2/StreamInterruptSuite.scala

```diff
@@ -369,4 +369,15 @@ class StreamInterruptSuite extends Fs2Suite {
     val interrupt = IO.sleep(250.millis)
     IO.race(compileWithSync(s).drain, interrupt).map(_.isRight).assert
   }
+
+  test("issue #2842 - no interrupt scope") {
+    val s = for {
+      local <- Stream.eval(IOLocal(List.empty[Int]))
+      _ <- Stream.eval(local.update(1 :: _)).noInterruptScope
+      _ <- Stream.eval(local.update(2 :: _)) // this one will be lost on a forked fiber
```

Comment:
yeah, that works. Going to your example in https://github.com/armanbilge/bayou/issues/1#issuecomment-1062944166 it would be like doing this I think:
```diff
- x  <- (spanS[IO]("stream") >> s2).interruptWhen(d).compile.foldMonoid
+ x  <- (spanS[IO]("stream") >> s2.interruptScope).interruptWhen(d).compile.foldMonoid
```

Reply:
Well, actually I'm not sure. Because even if s2 is interruptible I don't think it would respect the interruptibility of the outer `Stream` since it doesn't know about that.

Reply:
I think I understand your comment on the outer stream.  We don't want to go noInterrupt-interrupt.  We want to go noInterrupt-reset.  My `spanS` above would make an uninterruptible stream interruptible to trace it.

Reply:
Yeah, actually it doesn't matter at all whether the `Stream` we want to trace is interruptible or not. Really, we shouldn't be messing with that. In fact, interruptibility is sort of convoluted with the real problem here.

All we want is a way to call `local.set(...)` on the fiber will parent every other forked fiber in that `Stream`. (And ideally a way to do it again, at the end of the `Stream` when we close the `Span`). Because interruptibility is using `race` is why it becomes important in this context, but these are brittle sort of implementation details. And there could be other similar gotchas throughout fs2.

Since `IOLocal` is specific to `IO`, I wonder if we just need to add some `IO` specific methods to `Stream` specifically for working with locals. Because currently compositionality is kind of broken for anything to do with `IOLocal`.
```scala
val s = Stream.eval(local.set(...)) >> Stream.eval(local.get)
s.compile.lastOrError =!= s.interruptScope.compile.lastOrError
```

Reply:
Lazy question: does fs2 even depend on `IO`?  Wouldn't it need to put some sort of hint into the algebra that an `IO` interpreter was aware of?

Reply:
fs2 depends on CE core because of `SyncIO` but there's no direct dependency to `IO` that I'm aware of anyway.

I don't think that the interpreter needs to be aware of `IO`. Roughly I was thinking we could have some (ugly) method like this:

```scala
object Stream {
  def locally[A, B](local: IOLocal[A])(IOLocal[A] => IO[B]): Stream[IO, B]
}
```

that must satisfy something like

```scala
val s = Stream.locally(local)(_.set(...)) >> Stream.locally(local)(_.get)
s.compile.lastOrError === s.interruptScope.compile.lastOrError
```

The implementation can delegate to something like `Stream.evalUninterruptible` which can be a `private[fs2]` method. So that way we don't expose it, but also the algebra doesn't need to know about `IO`. The fact that the desired semantics are achieved by disabling interruption is remains an implementation detail.

Reply:
That sounds reasonable,  but it also means we can't derive `spanS` from the proposed `TraceResource` and would need a third type class:

```mermaid
classDiagram
    Trace <|-- TraceResource
    TraceResource <|-- TraceStream
      class Trace{
        +span(String name)(F~A~ fa) F~A~
      }
      class TraceResource{
        +spanR(String name) Resource~F~
      }
      class TraceStream{
        +spanS(String name) Stream~F~
      }
```

Sorry, I don't know enough Mermaid UML™ to render the output type of `Resource` and `Stream`, but you get the gist.

Or maybe `TraceResource` could have an abstract type parameter, and fs2 could have a `Locally` type class that relates `IO` and `IOLocal`, but now I'm feeling even dizzier.

Reply:
Yes. This is what I was getting at in https://github.com/armanbilge/bayou/issues/1#issuecomment-1062892012 and https://github.com/tpolecat/natchez/pull/526#issuecomment-1063614556: we will need a `spanS` method implemented specifically for `Stream` and thus would need to take on an `fs2` dependency. This would be true even with my original `noInterruptScope` idea.

I don't know if `TraceResource` and `TraceStream` or more typeclasses are the right answer 😅 starts to get complicated. I'd rather have an enhanced `Trace` and take on the fs2 dependency which is kind of what Bayou purported anyway.